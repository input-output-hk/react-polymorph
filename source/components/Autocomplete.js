// @flow
import React, { Component } from 'react';
import type { ComponentType, Element } from 'react';

// external libraries
import _ from 'lodash';

// interal components
import { GlobalListeners } from './HOC/GlobalListeners';

// internal utility functions
import { createEmptyContext, withTheme } from './HOC/withTheme';
import { composeTheme, addThemeId, didThemePropsChange } from '../utils/themes';
import { composeFunctions } from '../utils/props';

import { IDENTIFIERS } from '.';
import type { ThemeContextProp } from './HOC/withTheme';

export type AutocompleteProps = {
  className?: string,
  context: ThemeContextProp,
  error: ?string,
  invalidCharsRegex: RegExp,
  isOpeningUpward: boolean,
  label?: string | Element<any>,
  maxSelections?: number,
  requiredSelections: [number],
  requiredSelectionsInfo?: (required: number, actual: number) => string,
  maxVisibleOptions: number,
  multipleSameSelections: boolean,
  onChange?: Function,
  options: Array<any>,
  preselectedOptions?: Array<any>,
  placeholder?: string,
  renderSelections?: Function,
  renderOptions?: Function,
  skin?: ComponentType<any>,
  sortAlphabetically: boolean,
  theme: ?Object, // will take precedence over theme in context if passed
  themeId: string,
  themeOverrides: Object,
};

type State = {
  composedTheme: Object,
  error: string,
  filteredOptions: Array<any>,
  isOpen: boolean,
  inputValue: string,
  mouseIsOverOptions: boolean,
  selectedOptions: Array<any>,
  isRemoveWordClicked: boolean,
};

class AutocompleteBase extends Component<AutocompleteProps, State> {
  // declare ref types
  rootElement: ?Element<any>;
  inputElement: ?Element<'input'>;
  suggestionsElement: ?Element<any>;
  optionsElement: ?Element<any>;

  // define static properties
  static displayName = 'Autocomplete';
  static defaultProps = {
    context: createEmptyContext(),
    error: null,
    invalidCharsRegex: /[^a-zA-Z0-9\s]/g, // only allow letters and numbers by default
    isOpeningUpward: false,
    maxVisibleOptions: 10, // max number of visible options
    multipleSameSelections: true, // if true then same word can be selected multiple times
    options: [],
    requiredSelections: [],
    sortAlphabetically: true, // options are sorted alphabetically by default
    theme: null,
    themeId: IDENTIFIERS.AUTOCOMPLETE,
    themeOverrides: {},
  };

  constructor(props: AutocompleteProps) {
    super(props);

    // define refs
    this.rootElement = React.createRef();
    this.inputElement = React.createRef();
    this.suggestionsElement = React.createRef();
    this.optionsElement = React.createRef();

    const {
      context,
      themeId,
      theme,
      themeOverrides,
      sortAlphabetically,
      options,
      preselectedOptions,
    } = props;

    this.state = {
      inputValue: '',
      error: '',
      selectedOptions: preselectedOptions || [],
      filteredOptions:
        sortAlphabetically && options ? options.sort() : options || [],
      isOpen: false,
      isRemoveWordClicked: false,
      mouseIsOverOptions: false,
      composedTheme: composeTheme(
        addThemeId(theme || context.theme, themeId),
        addThemeId(themeOverrides, themeId),
        context.ROOT_THEME_API
      ),
    };
  }

  componentDidUpdate(prevProps: AutocompleteProps) {
    if (prevProps !== this.props) {
      didThemePropsChange(prevProps, this.props, this.setState.bind(this));
    }
  }

  clear = () => this._removeOptions();

  focus = () => {
    const { inputElement } = this;
    if (inputElement && inputElement.current) {
      inputElement.current.focus();
    }
  };

  open = () => this.setState({ isOpen: true });

  close = () => this.setState({ isOpen: false });

  toggleOpen = () => {
    if (
      this.state.isOpen &&
      this.optionsElement &&
      this.optionsElement.current
    ) {
      // set Options scroll position to top on close
      this.optionsElement.current.scrollTop = 0;
    }
    this.setState((prevState) => ({ isOpen: !prevState.isOpen }));
  };

  toggleMouseLocation = () =>
    this.setState((prevState) => ({
      mouseIsOverOptions: !prevState.mouseIsOverOptions,
    }));

  handleAutocompleteClick = () => {
    this.focus();
  };

  onKeyDown = (event: SyntheticKeyboardEvent<>) => {
    if (event) {
      if (
        // Check for backspace in order to delete the last selected option
        event.keyCode === 8 &&
        !event.target.value &&
        this.state.selectedOptions.length
      ) {
        // Remove last selected option
        this.close();
        this.removeOption(this.state.selectedOptions.length - 1);
      } else if (event.keyCode === 27) {
        // ESCAPE key: Stops propagation & modal closing
        event.stopPropagation();
      } else if (event.keyCode === 13) {
        // ENTER key: Opens suggestions
        this.open();
      }
    }
  };

  // onChange handler for input element in AutocompleteSkin
  handleInputChange = (event: SyntheticInputEvent<HTMLInputElement>) => {
    const { value } = event.target;
    const multipleValues = value.split(' ');
    const hasMultipleValues = multipleValues.length > 1;
    this._setInputValue(value);
    if (hasMultipleValues) {
      this.open();
    }
    this.updateSelectedOptions(multipleValues, multipleValues.length === 1);
  };

  // passed to Options onChange handler in AutocompleteSkin
  handleChange = (option: any) => {
    this.updateSelectedOptions(option);
  };

  updateSelectedOptions = (
    addedOption: string | Array<string>,
    singleInput?: boolean
  ) => {
    const { maxSelections, multipleSameSelections, options } = this.props;
    const { selectedOptions, isOpen } = this.state;
    const { filteredOptions } = this.state;
    const canMoreOptionsBeSelected =
      maxSelections != null ? selectedOptions.length < maxSelections : true;
    const areFilteredOptionsAvailable =
      filteredOptions && filteredOptions.length > 0;
    let skipValueSelection = false;
    if (
      !maxSelections ||
      (canMoreOptionsBeSelected && areFilteredOptionsAvailable)
    ) {
      if (!addedOption || !addedOption.length) return;
      const option = _.isString(addedOption)
        ? addedOption.trim()
        : addedOption.filter((item) => item !== '');
      const newSelectedOptions: Array<string> = [...selectedOptions];

      if (option && Array.isArray(option)) {
        option.forEach((item) => {
          const optionCanBeSelected =
            (multipleSameSelections &&
              options.includes(item) &&
              !singleInput &&
              option.length > 1) ||
            (options.includes(item) &&
              !selectedOptions.includes(item) &&
              !newSelectedOptions.includes(item) &&
              option.length > 1);
          if (
            (!optionCanBeSelected && !skipValueSelection) ||
            (singleInput && !selectedOptions.length)
          ) {
            this._setInputValue(item, true);
            skipValueSelection = true;
            return;
          }
          if (
            item &&
            optionCanBeSelected &&
            isOpen &&
            !skipValueSelection &&
            newSelectedOptions.length < maxSelections
          ) {
            newSelectedOptions.push(item);
          }
        });
      } else {
        const optionCanBeSelected =
          multipleSameSelections || !selectedOptions.includes(option);
        if (option && optionCanBeSelected && isOpen) {
          newSelectedOptions.push(option);
          skipValueSelection = false;
        }
      }
      this.selectionChanged(newSelectedOptions);
      this.setState({ selectedOptions: newSelectedOptions });
    } else {
      skipValueSelection = true;
    }
    if (!skipValueSelection) {
      this._setInputValue('');
    }
  };

  removeOption = (index: number) => {
    console.log(index);
    const { selectedOptions } = this.state;
    _.pullAt(selectedOptions, index);
    this.close();
    this.selectionChanged(selectedOptions);
    this.setState({ selectedOptions });
    this.focus();
  };

  selectionChanged = (selectedOptions: Array<any>) => {
    if (this.props.onChange) this.props.onChange(selectedOptions);
  };

  // returns an object containing props, theme, and method handlers
  // associated with rendering this.state.selectedOptions, the user can call
  // this in the body of the renderSelections function
  getSelectionProps = ({
    removeSelection,
  }: { removeSelection: Function } = {}) => {
    const { themeId } = this.props;
    const { inputValue, isOpen, selectedOptions, composedTheme } = this.state;
    return {
      inputValue,
      isOpen,
      selectedOptions,
      theme: composedTheme[themeId],
      removeSelection: (index: number) =>
        // the user's custom removeSelection event handler is composed with
        // the internal functionality of Autocomplete (this.removeOption)
        composeFunctions(removeSelection, this.removeOption)(index),
    };
  };

  render() {
    // destructuring props ensures only the "...rest" get passed down
    const {
      context,
      invalidCharsRegex,
      multipleSameSelections,
      sortAlphabetically,
      skin,
      theme,
      themeOverrides,
      onChange,
      error,
      ...rest
    } = this.props;

    const AutocompleteSkin = skin || context.skins[IDENTIFIERS.AUTOCOMPLETE];

    return (
      <GlobalListeners
        mouseIsOverOptions={this.state.mouseIsOverOptions}
        optionsIsOpen={this.state.isOpen}
        optionsIsOpeningUpward={this.props.isOpeningUpward}
        optionsRef={this.optionsElement}
        rootRef={this.rootElement}
        toggleOpen={this.toggleOpen}
      >
        {({ optionsMaxHeight, optionHeight }) => (
          <AutocompleteSkin
            error={error || this.state.error}
            filteredOptions={this.state.filteredOptions}
            getSelectionProps={this.getSelectionProps}
            handleAutocompleteClick={this.handleAutocompleteClick}
            handleChange={this.handleChange}
            handleInputChange={this.handleInputChange}
            inputRef={this.inputElement}
            inputValue={this.state.inputValue}
            isOpen={this.state.isOpen}
            onKeyDown={this.onKeyDown}
            optionsMaxHeight={optionsMaxHeight}
            optionsRef={this.optionsElement}
            removeOption={this.removeOption}
            rootRef={this.rootElement}
            selectedOptions={this.state.selectedOptions}
            suggestionsRef={this.suggestionsElement}
            theme={this.state.composedTheme}
            toggleMouseLocation={this.toggleMouseLocation}
            toggleOpen={this.toggleOpen}
            optionHeight={optionHeight}
            {...rest}
          />
        )}
      </GlobalListeners>
    );
  }

  // ======== PRIVATE METHOD ==========

  _removeOptions = () => {
    const { onChange } = this.props;
    onChange ? onChange([]) : null;
    this.setState({ selectedOptions: [], inputValue: '' });
  };

  _filterOptions = (value: string) => {
    let filteredOptions = [];

    if (value !== '') {
      _.some(this.props.options, (option) => {
        if (_.startsWith(option, value)) {
          filteredOptions.push(option);
        }
      });
    } else {
      filteredOptions = this.props.options;
    }

    return filteredOptions;
  };

  _filterInvalidChars = (value: string) => {
    let filteredValue = '';

    if (this.props.invalidCharsRegex.test(value)) {
      filteredValue = value.replace(this.props.invalidCharsRegex, '');
    } else {
      filteredValue = value;
    }

    return filteredValue;
  };

  _setInputValue = (value: string, shouldFocus?: boolean) => {
    const multipleValues = value.split(' ');
    if (multipleValues && multipleValues.length > 1) {
      let selectedOptions = [];
      multipleValues.forEach((itemValue) => {
        const filteredValue = this._filterInvalidChars(itemValue);
        selectedOptions = [
          ...selectedOptions,
          ...this._filterOptions(filteredValue),
        ];
      });
      this.setState({
        isOpen: !!multipleValues.length,
        inputValue: '',
        filteredOptions: Array.from(new Set(selectedOptions)),
      });
    } else {
      const filteredValue = this._filterInvalidChars(value);
      const filteredOptions = this._filterOptions(filteredValue);
      this.setState({
        isOpen: value === '' ? false : !!value,
        inputValue: filteredValue,
        filteredOptions,
      });
      setTimeout(() => {
        if (shouldFocus) this.focus();
      }, 0);
    }
  };
}

export const Autocomplete = withTheme(AutocompleteBase);
