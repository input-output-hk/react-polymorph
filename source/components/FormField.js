// @flow
import React, { Component } from 'react';
import type { ElementRef, ComponentType, Element } from 'react';

// internal utility functions
import { createEmptyContext, withTheme } from './HOC/withTheme';
import { composeTheme, addThemeId, didThemePropsChange } from '../utils/themes';

// import constants
import { IDENTIFIERS } from '.';
import type { ThemeContextProp } from './HOC/withTheme';

export type FormFieldProps = {
  className?: ?string,
  context: ThemeContextProp,
  disabled?: boolean,
  error?: string | Element<any>,
  formFieldRef: ElementRef<*>,
  id?: string,
  isErrorHidden?: boolean,
  isErrorShown?: boolean,
  isShowingErrorOnFocus: boolean,
  isShowingErrorOnHover: boolean,
  label?: string | Element<any>,
  onChange: Function,
  render: (setFormFieldRef: (ElementRef<*>) => void) => React$Node,
  skin?: ComponentType<any>,
  theme: ?Object, // will take precedence over theme in context if passed
  themeId: string,
  themeOverrides: Object,
  themeVariables?: Object,
};

type State = {
  error: string,
  composedTheme: Object,
};

class FormFieldBase extends Component<FormFieldProps, State> {
  // define static properties
  static displayName = 'FormField';

  static defaultProps = {
    context: createEmptyContext(),
    isShowingErrorOnFocus: true,
    isShowingErrorOnHover: true,
    theme: null,
    themeId: IDENTIFIERS.FORM_FIELD,
    themeOverrides: {},
  };

  formFieldRef: ElementRef<*>;

  constructor(props: FormFieldProps) {
    super(props);

    const { context, themeId, theme, themeOverrides } = props;

    this.formFieldRef = props.formFieldRef ?? React.createRef();

    this.state = {
      error: '',
      composedTheme: composeTheme(
        addThemeId(theme || context.theme, themeId),
        addThemeId(themeOverrides, themeId),
        context.ROOT_THEME_API
      ),
    };
  }

  componentDidUpdate(prevProps: FormFieldProps) {
    if (prevProps !== this.props) {
      didThemePropsChange(prevProps, this.props, this.setState.bind(this));
    }
  }

  setError = (error: string) => this.setState({ error });

  focusChild = () => {
    const { formFieldRef } = this;
    if (formFieldRef && formFieldRef.current) {
      if (typeof formFieldRef.current.focus === 'function') {
        formFieldRef.current.focus();
      }
    }
  };

  render() {
    // destructuring props ensures only the "...rest" get passed down
    const { skin, theme, themeOverrides, error, context, ...rest } = this.props;

    const FormFieldSkin = skin || context.skins[IDENTIFIERS.FORM_FIELD];

    return (
      <FormFieldSkin
        error={error || this.state.error}
        setError={this.setError}
        theme={this.state.composedTheme}
        formFieldRef={this.formFieldRef}
        focusChild={this.focusChild}
        {...rest}
      />
    );
  }
}

export const FormField = withTheme(FormFieldBase);
