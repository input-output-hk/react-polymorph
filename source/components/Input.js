// @flow
import React, { Component, RefObject } from 'react';
// $FlowFixMe
import type { ComponentType, Element, SyntheticInputEvent } from 'react';

// external libraries
import { isString, flow } from 'lodash';

// utilities
import { createEmptyContext, withTheme } from './HOC/withTheme';
import { composeTheme, addThemeId, didThemePropsChange } from '../utils/themes';

// constants
import { IDENTIFIERS } from '.';
import type { ThemeContextProp } from './HOC/withTheme';

export type InputProps = {
  autoFocus: boolean,
  className?: ?string,
  context: ThemeContextProp,
  disabled?: boolean,
  error?: string | Element<any>,
  inputRef?: RefObject,
  showErrorState?: boolean,
  hideErrorState?: boolean,
  id?: string,
  isShowingErrorOnFocus: boolean,
  isShowingErrorOnHover: boolean,
  label?: string | Element<any>,
  maxLength?: number,
  minLength?: number,
  onChange?: Function,
  placeholder?: string,
  readOnly: boolean,
  setError?: Function,
  selectedOption?: any,
  selectionRenderer?: Function,
  skin?: ComponentType<any>,
  theme: ?Object, // will take precedence over theme in context if passed
  themeId: string,
  themeOverrides: Object,
  themeVariables?: Object,
  value: string,
};

type State = {
  error: string,
  composedTheme: Object,
};

class InputBase extends Component<InputProps, State> {
  inputElement: RefObject;

  static displayName = 'Input';

  static defaultProps = {
    autoFocus: false,
    context: createEmptyContext(),
    error: '',
    isShowingErrorOnFocus: true,
    isShowingErrorOnHover: true,
    readOnly: false,
    theme: null,
    themeId: IDENTIFIERS.INPUT,
    themeOverrides: {},
    value: '',
  };

  constructor(props: InputProps) {
    super(props);
    this.inputElement = props.inputRef ?? React.createRef();
    const { context, themeId, theme, themeOverrides } = props;

    this.state = {
      composedTheme: composeTheme(
        addThemeId(theme || context.theme, themeId),
        addThemeId(themeOverrides, themeId),
        context.ROOT_THEME_API
      ),
      error: '',
    };
  }

  componentDidMount() {
    if (this.props.autoFocus) this.focus();
  }

  componentDidUpdate(prevProps: Props) {
    if (prevProps !== this.props) {
      didThemePropsChange(prevProps, this.props, this.setState.bind(this));
    }
  }

  onChange = (event: SyntheticInputEvent<Element<'input'>>) => {
    const { onChange, disabled, readOnly } = this.props;
    if (disabled || readOnly) return;
    if (onChange) onChange(this._processValue(event.target.value), event);
  };

  focus = () => {
    const { inputElement } = this;
    if (!inputElement.current) return;
    inputElement.current.focus();
  };

  _setError = (error: string) => {
    const { setError } = this.props;

    // checks for setError func from FormField component
    // if this Input instance is being used within the render function
    // of a FormField instance, the error field within FormField's local state
    // will be set
    if (setError) setError(error);
    this.setState({ error });
  };

  _processValue(value: string) {
    return flow([
      this._enforceStringValue,
      this._enforceMaxLength,
      this._enforceMinLength,
    ]).call(this, value);
  }

  _enforceStringValue(value) {
    if (!isString(value)) {
      throw new Error('Values passed to Input::onChange must be strings');
    }
    return value;
  }

  _enforceMaxLength(value: string) {
    const { maxLength } = this.props;
    const isTooLong = maxLength != null && value.length > maxLength;
    return isTooLong ? value.substring(0, maxLength) : value;
  }

  _enforceMinLength(value: string) {
    const { minLength } = this.props;
    const isTooShort = minLength != null && value.length < minLength;

    if (isTooShort) {
      this._setError('Please enter a valid input');
    } else if (this.state.error !== '') {
      this._setError('');
    }

    return value;
  }

  render() {
    // destructuring props ensures only the "...rest" get passed down
    const {
      skin,
      context,
      theme,
      themeOverrides,
      onChange,
      error,
      maxLength,
      minLength,
      setError,
      autoFocus,
      ...rest
    } = this.props;

    const InputSkin = skin || context.skins[IDENTIFIERS.INPUT];

    return (
      <InputSkin
        error={error || this.state.error}
        onChange={this.onChange}
        inputRef={this.inputElement}
        theme={this.state.composedTheme}
        {...rest}
      />
    );
  }
}

export const Input = withTheme(InputBase);
