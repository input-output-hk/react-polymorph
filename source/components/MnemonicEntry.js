// @flow
import React, { Component } from 'react';
import type { ComponentType, Element } from 'react';

// internal utility functions
import { createEmptyContext, withTheme } from './HOC/withTheme';
import { composeTheme, addThemeId, didThemePropsChange } from '../utils/themes';

// import constants
import { IDENTIFIERS } from '.';
import type { ThemeContextProp } from './HOC/withTheme';

type Props = {
  className?: string,
  context: ThemeContextProp,
  displayOnly: boolean,
  label?: string | Element<any>,
  mnemonicWords: Array<string>,
  skin?: ComponentType<any>,
  theme: ?Object, // will take precedence over theme in context if passed
  themeId: string,
  themeOverrides: Object, // custom css/scss from user that adheres to component's theme API
  totalColumns: number,
};

type State = {
  activeColumn: number | null,
  composedTheme: Object,
  totalWordsEntered: number,
};

class MnemonicEntryBase extends Component<Props, State> {
  // define static properties
  static displayName = 'MnemonicEntry';
  static defaultProps = {
    context: createEmptyContext(),
    displayOnly: true,
    loading: false,
    theme: null,
    themeId: IDENTIFIERS.MNEMONIC_ENTRY,
    themeOverrides: {}
  };

  constructor(props: Props) {
    super(props);

    const { context, themeId, theme, themeOverrides } = props;

    this.state = {
      activeColumn: null,
      composedTheme: composeTheme(
        addThemeId(theme || context.theme, themeId),
        addThemeId(themeOverrides, themeId),
        context.ROOT_THEME_API
      ),
      totalWordsEntered: 0,
    };
  }

  componentWillMount() {
    const { totalColumns } = this.props;

    if (document.documentElement && document.documentElement.style) {
      document.documentElement.style.setProperty('--rp-mnemonicentry-number-of-columns', `${totalColumns}`);
    }
  }

  componentWillReceiveProps(nextProps: Props) {
    didThemePropsChange(this.props, nextProps, this.setState.bind(this));
  }

  setActiveColumn = (columnNumber: number) => this.setState({ activeColumn: columnNumber });

  transformWordsToObj = () => {
    const mnemonicWordsObj = {};
    this.props.mnemonicWords.forEach((word, index) => mnemonicWordsObj[`${index + 1}`] = word);
    return mnemonicWordsObj;
  }

  render() {
    // destructuring props ensures only the "...rest" get passed down
    const {
      skin,
      theme,
      themeOverrides,
      context,
      mnemonicWords,
      ...rest
    } = this.props;

    const {
     activeColumn,
     composedTheme,
     totalWordsEntered,
    } = this.state;

    const MnemonicEntrySkin = skin || context.skins[IDENTIFIERS.MNEMONIC_ENTRY];
    const totalWords = mnemonicWords.length;
    const mnemonicWordsObj = this.transformWordsToObj();

    return (
      <MnemonicEntrySkin
        activeColumn={activeColumn}
        mnemonicWords={mnemonicWordsObj}
        setActiveColumn={this.setActiveColumn}
        theme={composedTheme}
        totalWords={totalWords}
        totalWordsEntered={totalWordsEntered}
        {...rest}
      />
    );
  }
}

export const MnemonicEntry = withTheme(MnemonicEntryBase);
