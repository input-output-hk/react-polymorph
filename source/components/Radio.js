// @flow
import React, { Component } from 'react';
import type { ComponentType, Element } from 'react';

// internal utility functions
import { createEmptyContext, withTheme } from './HOC/withTheme';
import { composeTheme, addThemeId, didThemePropsChange } from '../utils/themes';

// import constants
import { IDENTIFIERS } from '.';
import type { ThemeContextProp } from './HOC/withTheme';

type Props = {
  context: ThemeContextProp,
  disabled?: boolean,
  label?: string | Element<any>,
  onBlur?: Function,
  onChange?: Function,
  onFocus?: Function,
  selected: boolean,
  skin?: ComponentType<any>,
  theme: ?Object, // will take precedence over theme in context if passed
  themeId: string,
  themeOverrides: Object
};

type State = {
  composedTheme: Object
};

class RadioBase extends Component<Props, State> {
  // define static properties
  static displayName = 'Radio';
  static defaultProps = {
    context: createEmptyContext(),
    selected: false,
    theme: null,
    themeId: IDENTIFIERS.RADIO,
    themeOverrides: {}
  };

  constructor(props: Props) {
    super(props);

    const { context, themeId, theme, themeOverrides } = props;

    this.state = {
      composedTheme: composeTheme(
        addThemeId(theme || context.theme, themeId),
        addThemeId(themeOverrides, themeId),
        context.ROOT_THEME_API
      )
    };
  }

  componentDidUpdate(prevProps: Props) {
    if (prevProps !== this.props) {
      didThemePropsChange(prevProps, this.props, this.setState.bind(this));
    }
  }

  render() {
    // destructuring props ensures only the "...rest" get passed down
    const { skin, theme, themeOverrides, context, ...rest } = this.props;

    const RadioSkin = skin || context.skins[IDENTIFIERS.RADIO];

    return <RadioSkin theme={this.state.composedTheme} {...rest} />;
  }
}

export const Radio = withTheme(RadioBase);
