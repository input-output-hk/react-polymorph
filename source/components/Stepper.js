// @flow
import React, { Component } from 'react';
import type { ComponentType } from 'react';

// internal utility functions
import { createEmptyContext, withTheme } from './HOC/withTheme';
import { composeTheme, addThemeId, didThemePropsChange } from '../utils/themes';

// import constants
import { IDENTIFIERS } from '.';
import type { ThemeContextProp } from './HOC/withTheme';

type Props = {
  activeStep?: number,
  className?: string,
  context: ThemeContextProp,
  label?: string,
  labelDisabled?: boolean,
  onStepClick?: Function,
  skin?: ComponentType<any>,
  steps: Array<string>,
  theme: ?Object, // will take precedence over theme in context if passed
  themeId: string,
  themeOverrides: Object
};

type State = {
  composedTheme: Object
};

class StepperBase extends Component<Props, State> {
  // define static properties
  static displayName = 'Stepper';
  static defaultProps = {
    context: createEmptyContext(),
    theme: null,
    themeId: IDENTIFIERS.STEPPER,
    themeOverrides: {}
  };

  constructor(props: Props) {
    super(props);

    const { context, themeId, theme, themeOverrides } = props;

    this.state = {
      composedTheme: composeTheme(
        addThemeId(theme || context.theme, themeId),
        addThemeId(themeOverrides, themeId),
        context.ROOT_THEME_API
      )
    };
  }

  componentDidUpdate(prevProps: Props) {
    if (prevProps !== this.props) {
      didThemePropsChange(prevProps, this.props, this.setState.bind(this));
    }
  }

  render() {
    // destructuring props ensures only the "...rest" get passed down
    const { skin, theme, themeOverrides, context, ...rest } = this.props;

    const StepperSkin = skin || context.skins[this.props.themeId];

    return <StepperSkin theme={this.state.composedTheme} {...rest} />;
  }
}

export const Stepper = withTheme(StepperBase);
