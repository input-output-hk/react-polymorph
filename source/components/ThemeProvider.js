// @flow
import React, { Component } from 'react';
import type { Node } from 'react';

// external libraries
import { isEmpty, isEqual, cloneDeep } from 'lodash';

// contains default theme and context provider
import { ThemeContext } from './HOC/ThemeContext';

// imports the Root Theme API object which specifies the shape
// of a complete theme for every component in this library, used in this.composeLibraryTheme
import { ROOT_THEME_API } from '../themes/API';

// internal utility functions
import { appendToProperty } from '../utils/themes';
import { hasProperty } from '../utils/props';
import { ThemeVariablesProvider } from './ThemeVariablesProvider';
import type { ThemeVariables } from './ThemeVariablesProvider';

type Props = {
  children?: ?Node,
  skins: Object,
  theme: Object,
  variables: ThemeVariables,
  isRoot: boolean,
  themeOverrides: Object, // custom css/scss from user that adheres to shape of ROOT_THEME_API
};

type State = {
  theme: Object,
};

export class ThemeProvider extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    const { theme, themeOverrides } = props;
    this.state = {
      theme: this._composeLibraryTheme(theme, themeOverrides),
    };
  }

  componentDidUpdate(prevProps: Props) {
    const { theme: prevTheme, themeOverrides: prevThemeOverrides } = prevProps;
    const { theme, themeOverrides } = this.props;

    if (
      !isEqual(prevTheme, theme) ||
      !isEqual(prevThemeOverrides, themeOverrides)
    ) {
      this.setState(() => ({
        theme: this._composeLibraryTheme(theme, themeOverrides),
      }));
    }
  }

  // composeLibraryTheme returns a single obj containing theme definitions
  // for every component in the library. Every key on the returned obj is named
  // in conjunction with a component in the library and each key's value is structured
  // to contain the css definitions for each element in that component.
  // Which is just a string via CSS-Modules. Looks like this:
  // {
  //   button: { root: '', disabled: '' },
  //   input: { input: '', disabled: '', error: '' },
  //   formField: { root: '', label: '', error: '' },
  //   ... and so on, creating a complete theme for the library,
  //  }
  _composeLibraryTheme = (theme: Object, themeOverrides: Object) => {
    // if themeOverrides is empty, no need for composition
    if (isEmpty(themeOverrides)) {
      return theme;
    }

    // final object to be returned
    const composedTheme = {};

    for (const componentName in ROOT_THEME_API) {
      // check if ROOT_THEME_API contains the key of componentName
      if (hasProperty(ROOT_THEME_API, componentName)) {
        // check if theme contains a key of componentName
        if (hasProperty(theme, componentName)) {
          // add componentName as a key to final return obj
          composedTheme[componentName] = theme[componentName];
        }

        // also check if themeOverrides contains the key componentName
        if (hasProperty(themeOverrides, componentName)) {
          // compose theme styles with user's themeOverrides
          composedTheme[componentName] = this._applyThemeOverrides(
            theme[componentName],
            themeOverrides[componentName],
            ROOT_THEME_API[componentName]
          );
        }
      }
    }
    return composedTheme;
  };

  _applyThemeOverrides = (
    componentTheme: Object,
    componentThemeOverrides: Object,
    componentThemeAPI: Object
  ) => {
    // Return componentTheme if there are no overrides provided
    if (isEmpty(componentThemeOverrides)) {
      return componentTheme;
    }

    // final composed theme obj to be returned at end
    const composedComponentTheme = cloneDeep(componentThemeAPI);

    for (const className in componentThemeAPI) {
      if (hasProperty(componentThemeAPI, className)) {
        if (hasProperty(componentTheme, className)) {
          appendToProperty(
            composedComponentTheme,
            className,
            componentTheme[className]
          );
        }

        if (hasProperty(componentThemeOverrides, className)) {
          appendToProperty(
            composedComponentTheme,
            className,
            componentThemeOverrides[className]
          );
        }
      }
    }
    return composedComponentTheme;
  };

  render() {
    const { theme } = this.state;
    const { isRoot, skins, variables } = this.props;
    const providerState = { skins, theme, ROOT_THEME_API };
    return (
      <ThemeContext.Provider value={providerState}>
        <ThemeVariablesProvider isRoot={isRoot} variables={variables}>
          {this.props.children}
        </ThemeVariablesProvider>
      </ThemeContext.Provider>
    );
  }
}

ThemeProvider.defaultProps = {
  isRoot: true,
  skins: {},
  theme: {},
  variables: {},
  themeOverrides: {},
};
