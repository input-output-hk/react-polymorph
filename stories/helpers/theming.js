import React from 'react';
import { ThemeProvider } from '../../source/components/ThemeProvider';
import { SimpleTheme, SimpleDefaults } from '../../source/themes/simple';
import { SimpleSkins } from '../../source/skins/simple';

export const decorateWithSimpleTheme = (story) => (
  <ThemeProvider
    skins={SimpleSkins}
    theme={SimpleTheme}
    variables={SimpleDefaults}
  >
    {story()}
  </ThemeProvider>
);
