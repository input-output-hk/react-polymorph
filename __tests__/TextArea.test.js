import React from 'react';

import { TextArea } from '../source/components/TextArea';
import { renderInSimpleTheme } from './helpers/theming';

test('TextArea renders correctly', () => {
  expect(renderInSimpleTheme(<TextArea />)).toMatchSnapshot();
});

test('TextArea renders with placeholder', () => {
  expect(
    renderInSimpleTheme(<TextArea placeholder="0.0000" />)
  ).toMatchSnapshot();
});

test('TextArea renders with a value', () => {
  expect(
    renderInSimpleTheme(<TextArea value="this one has a value" />)
  ).toMatchSnapshot();
});

test('TextArea renders with 5 rows', () => {
  expect(renderInSimpleTheme(<TextArea rows={5} />)).toMatchSnapshot();
});
