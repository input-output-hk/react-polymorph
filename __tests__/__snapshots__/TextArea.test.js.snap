// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea renders correctly 1`] = `
<div
  class="root"
>
  <span>
    <div
      class="inputWrapper"
    >
      <textarea
        class="textarea"
      />
    </div>
  </span>
</div>
`;

exports[`TextArea renders with 5 rows 1`] = `
<div
  class="root"
>
  <span>
    <div
      class="inputWrapper"
    >
      <textarea
        class="textarea"
        rows="5"
      />
    </div>
  </span>
</div>
`;

exports[`TextArea renders with a value 1`] = `
<div
  class="root"
>
  <span>
    <div
      class="inputWrapper"
    >
      <textarea
        class="textarea"
      >
        this one has a value
      </textarea>
    </div>
  </span>
</div>
`;

exports[`TextArea renders with placeholder 1`] = `
<div
  class="root"
>
  <span>
    <div
      class="inputWrapper"
    >
      <textarea
        class="textarea"
        placeholder="0.0000"
      />
    </div>
  </span>
</div>
`;
